runtime/oti/VMHelpers.hpp

```diff
@@ -2068,9 +2070,61 @@ class VM_VMHelpers
 	static VMINLINE bool
 	isContinuationMounted(J9VMContinuation *continuation)
 	{
-		return (NULL != continuation->carrierThread);
+		return J9_ARE_ANY_BITS_SET(continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+	}
+
+	static VMINLINE bool
+	isContinuationMountedOrConcurrentlyScanned(J9VMContinuation *continuation)
+	{
+		return isContinuationMounted(continuation) || isConcurrentlyScannedFromContinuationState(continuation->state);
 	}
+
+/*
+ *	If low tagging failed due to either
+ *
+ *   a carrier thread winning to mount, we don't need to do anything, since it will be compensated by pre/post mount actions
+ *   another GC thread winning to scan, again don't do anything, and let the winning thread do the work, instead
+ */
+	static VMINLINE bool
+	tryWinningConcurrentGCScan(J9VMContinuation *continuation)
+	{
+		return J9_GC_CONTINUATION_STATE_INITIAL == VM_AtomicSupport::lockCompareExchange(&continuation->state, J9_GC_CONTINUATION_STATE_INITIAL, J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+	}
+
+	static VMINLINE void
+	exitConcurrentGCScan(J9VMContinuation *continuation)
+	{
+		/* clear CONCURRENTSCANNING flag */
+		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+		J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
+		if (NULL != carrierThread) {
+			omrthread_monitor_enter(carrierThread->publicFlagsMutex);
+			/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
+			omrthread_monitor_notify(carrierThread->publicFlagsMutex);
```

Comment:
As a rule, we always notify all on publicFlagsMutex. This is especially important since you're adding new uses of the monitor.

I'm also concerned that you're not protecting anything with the monitor when notifying - the waiter is acquiring the monitor before reading the continuation state. I'm not convinced this is actually an error, but it breaks the well-used pattern from everywhere else.

Reply:
change `omrthread_monitor_notify` to `omrthread_monitor_notify_all`

Reply:
Agree, this must be notify_all.

The other concern is also valid, and yes, I agree it does look unusual that we don't set/reset a flag while holding monitor.

At one point we did have a new HALT flag for this, and GC thread would clear the halt flag and notify halted carrier all while holding monitor. Carrier would then unblock via VM access re-acquire (rather than relying on explicitly introduced loop).

But we envisioned problems with it. A specific example is: there would be a gap in time during mount, between a moment when carrier thread realizes there is GC in progress (by seeing tag in the state) and a moment when it sets the halt flag and starts waiting (on VM acquire). During that gap, GC could finish and clear the halt flag (even though it was not even set yet), and then carrier thread sets and goes waiting, what would continue on indefinitely, since notification was already gone. There was a disconnect between initial decision to wait (result from atomic on state) and how long it should wait (till halt flag is cleared). 

Should be really based on once piece of info, and as we have it now, carrier thread will be in an explicit waiting loop until the same tag from the state is cleared. The atomics on a single state variable are there to make sure both thread mounting and GC thread scanning agree and decisions are made coherently:
- carrier will enter wait loop only if during atomic it saw there is an ongoing GC scan
- GC is guarantied to notify waiting thread since during atomic while clearing the tag it saw there was indeed a carrier thread waiting (along the way it obtains a handle on its monitor).

Reply:
Since the update is done with an atomic, I think this will be fine.